File: examples/playground/features/higher-order.js
// Functions are first-class. Inty handles higher-order use directly.

function compose(f, g) {
    return function(x) { return f(g(x)); };
}

function double(x) { return x * 2; }
function inc(x)    { return x + 1; }

var f = compose(double, inc);
var n = f(3);                       // (3 + 1) * 2 = 8

// `compose` is fully polymorphic — works for any matching pair.
var g = compose(function(s) { return s + "!"; },
                function(s) { return s + s; });
var loud = g("yo");                 // "yoyo!"

File: examples/playground/py-features/higher-order.py
# Functions are first-class. `compose` is fully polymorphic — inty
# infers `compose<a, b, c>((b) => c, (a) => b) => ((a) => c)`.

def compose(f, g):
    def composed(x):
        return f(g(x))
    return composed

def double(n):
    return n * 2

def inc(n):
    return n + 1

f = compose(double, inc)
m = f(3)              # (3 + 1) * 2 == 8

# Same compose, different types — strings flow through fine.
def loud(s):
    return s + "!"
def repeat(s):
    return s + s

g = compose(loud, repeat)
hi = g("yo")          # "yoyo!"
